### PR TITLE
fix: Not all commits beeing evalutated, added all option to gitlog

### PR DESCRIPTION
**Pull request by Forestsoft-de**
In my case only merge commits are evalulated. I dont know why but get-next-version always decide that there are only chore commits.

After adding All: true it works.

BR
Sebastian

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -6,3 +6,4 @@ coverage/
 
 # OS generated files and directories
 .DS_Store
+get-next-version
```
Comment (goloroden) [inline .gitignore:9]: This shouldn't be needed, if the build is run using the `Makefile`, with the command `make build`. Can you please revert this change?

File: git/commits.go [hunk 2]
```diff
@@ -29,7 +29,7 @@ func GetConventionalCommitTypesSinceLastRelease(repository *git.Repository) (Con
 		}
 		return ConventionalCommmitTypesResult{}, err
 	}
-	commitIterator, err := repository.Log(&git.LogOptions{From: head.Hash()})
+	commitIterator, err := repository.Log(&git.LogOptions{From: head.Hash(), All: true, Order: git.LogOrderCommitterTime})
```
Comment (goloroden) [inline git/commits.go:32]: According to the [documentation of the `go-git` module](https://pkg.go.dev/github.com/go-git/go-git/v5#LogOptions) using `From` and `Order` at the same time is incompatible. I.e., if `Order` is being used, `From` is ignored. So we should remove one of these. If it's actually needed, we should remove `From`. If it is not needed, we should remove it.

Can you please double-check this?

**Review by goloroden — CHANGES_REQUESTED** [review]
Hey @Forestsoft-de 👋 

First of all: Thank you very much for your PRs, we appreciate this a lot 😊

I've decided to first focus on this one, as this is the easier one, and the other one needs a little bit more of discussion. I hope this is fine for you.

I've reviewed your changes, and left a few comments. Could you please take care of them?

In addition it would be great if we could have a test that reproduces the faulty behavior. Can you please add one?

Have a nice day,


Golo

**Comment by Forestsoft-de** [comment]
> Hey @Forestsoft-de 👋
> 
> First of all: Thank you very much for your PRs, we appreciate this a lot 😊
> 
> I've decided to first focus on this one, as this is the easier one, and the other one needs a little bit more of discussion. I hope this is fine for you.

I also guess that it could rely on different versions between Debian and alpline git but it didnt. The behavior was reproduceable with the debian container.

> I've reviewed your changes, and left a few comments. Could you please take care of them?
> 
> In addition it would be great if we could have a test that reproduces the faulty behavior. Can you please add one?

Yeah sure but you know I am the PHP guy :D 
I ran an debugging session with my repo and i detected that the repository.Log(&git.LogOptions{From: head.Hash()}) returns only merged commits. I saw that one of the merge commit has also a tag:

For example
git log --oneline 

d623a86 (HEAD -> main, origin/main, origin/HEAD) Merge pull request #10 from ForestsoftGmbH/feature/manual_deployment
6985e3a (origin/feature/manual_deployment) fix: make it possible to redeploy an exisisting release
eb2d6b7 Trigger pipeline manually for new version
e867e66 (tag: 0.1.15) Merge pull request #9 from ForestsoftGmbH/bugfix/newversion

In this case only d623a86 and e867e66 are evaluated in the GetConventionalCommitTypesSinceLastRelease function.

Unfortunately i didnt know how to reproduce this behavior in the testcase because you define the history as you expect :)


> Have a nice day,
> 
> Golo

**Comment by goloroden** [comment]
To somehow move this forward – could you maybe provide a repository that we can use to reproduce the behavior?

Right now I'm not sure how to move on from the current status quo, since we can't reproduce things, and can not exactly tell which effect it will have if we merge this.

**Comment by Forestsoft-de** [comment]
@goloroden sure. I setuped an test repository and implemented the same workflow as i use for my private projects:

https://github.com/ForestsoftGmbH/tnw-semver-tester/actions/runs/3295316981/jobs/5433820444

In this action i would expect the versio 0.2.1 because we have one commit with feat: and one with fix.

The result is 0.1.0 as it would before in the tagged commit on main branch.

BR
Sebastian

**Comment by goloroden** [comment]
Hi @Forestsoft-de 👋 

Thanks a lot for setting up the example repository.

Actually, the current behavior is correct:

- Since no tag was given in the beginning, `get-next-version` starts with an assumed version of `0.0.0`.
- Then you merged a `feat` and a `fix`, which basically means `minor` and `patch`.
- Since minor are allowed to include patches, the `fix` is not relevant here, only the `feat` is considered.
- Hence the minor version is increased, and you get a `0.1.0`.

So, it behaves as expected (and conforming to the way Semver works).

**Comment by goloroden** [comment]
PS: Since the software works as designed, I'm going to close this PR.

**Comment by Forestsoft-de** [comment]
Check the gitlog. There is an tag on a previous commit in main branch. 

And i saw it in the debugger that some commits are not handled by the get-next-version code. That was the intention of this PR.

BR
Sebastian

**Comment by goloroden** [comment]
The `0.1.0` tag is on the third commit, and after that there is only the merge commit, which is missing a prefix, hence no new version is being generated.

**Comment by goloroden** [comment]
PS: Somehow I have the feeling that we are at cross purposes … from what I am able to understand, the history is as follows:

- Initlal commit
- fix commit
- feat commit
- Tag 0.1.0
- Merge commit

Was the tag created manually, or using `get-next-version`?

And, at what point in time do you now try to run `get-next-version` and expect the `0.2.0`?

**Comment by goloroden** [comment]
Because, up to `0.1.0` there was a `feat` and a `fix`, and no previous tag, so `0.1.0` is correct.

And after the `0.1.0` tag there are no commits with any prefix, so they are ignored.

**Comment by goloroden** [comment]
Okay, thanks for clarifying this 😊

My first approach was to clone your repository, switch to the branch you mentioned, and run `get-next-version` locally. As expected, the result is `0.2.0`:

```shell
$ git clone git@github.com:ForestsoftGmbH/tnw-semver-tester.git
$ git checkout fix/deprecated-warning
$ get-next-version
0.2.0
```

So, apparently the current version of the tool performs the right thing.

The question is: Why does it behave differently on GitHub?

**Comment by Forestsoft-de** [comment]
@goloroden you should check it with the docker container you provide in the action. Maybe there is something different.

As i mentionied with my patch it works as expected. I already use it in multiple of my pipelines with an own fixed docker container:

https://hub.docker.com/repository/docker/forestsoft/get-next-version

BR
Sebastian

**Comment by goloroden** [comment]
I agree that there is something wrong, but if it works correctly locally, the issue is not in the code of the tool, but rather inside how we build / use the Docker image, is it?

So, I'm not sold yet on the code change, because apparently it also works correctly without this change, just not in the Dockerized situation.

And we need to figure out what causes the problem there.

**Comment by goloroden** [comment]
The bug is also not within the Docker image.

I built the Docker image locally, and ran it against your repository, and it also produces `0.2.0`.
Hmmm 🤔

The only thing that remains then is the way the workflow is written that uses `get-next-version`.

According to our documentation, you did everything correctly (especially set the `fetch-depth` parameter), but we have only ever tested this on the `main` branch.

My current guess is that (for somewhat reason) your workflow is running on the wrong branch … but that's just a gut feeling, and I can not yet explain it. I'll get back to you once I have new insights.

**Comment by goloroden** [comment]
Okay, I've got the solution (I guess): https://github.com/actions/checkout/issues/648

When using `fetch-depth: 0` the current branch is ignored, and you *always* get the `main` branch.
You need to specify the desired branch / commit by providing `ref: <...>`.

As said, this did never catch our eye before since we only ever run this on the `main` branch.

Can you please try to provide the correct `ref`, and see if this fixes the issue?

**Comment by Forestsoft-de** [comment]
Hey @goloroden 

the ref: ${{ github.event.pull_request.head.sha }} results in:

{"level":"fatal","time":1666449201,"message":"object not found"}

See: https://github.com/ForestsoftGmbH/tnw-semver-tester/actions/runs/3303603725/jobs/5451727442

BR
Sebastian

**Comment by goloroden** [comment]
I've opened a PR on your repository by myself, to make some tests – could you maybe allow it to run? See https://github.com/ForestsoftGmbH/tnw-semver-tester/pull/3

**Comment by Forestsoft-de** [comment]
@goloroden it works now.

Ive update the PR to only improved example in your readme.

BR
Sebastian

**Review by goloroden — APPROVED** [review]

**Pull request merged**